File: Introduction_to_Robotics_Course/Chapter7/Lesson1/Chapter7_Lesson1.py
import numpy as np

# 1D robot: x=[p, v]^T
Ts = 0.1
A = np.array([[1, Ts],
              [0, 1]])
B = np.array([[0.5*Ts**2],
              [Ts]])

# true state
x = np.array([[0.0],
              [1.0]])

# proprio velocity sensor bias
b = 0.05
sigma_p = 0.02
sigma_e = 0.05

p_hat_from_proprio = 0.0

N = 200
p_true_hist, p_hat_hist, p_extero_hist = [], [], []

for k in range(N):
    u = np.array([[0.0]])  # constant velocity case
    w = np.zeros((2,1))
    x = A @ x + B @ u + w

    # proprio sensor: velocity with bias
    y_p = x[1,0] + b + np.random.randn()*sigma_p
    p_hat_from_proprio += Ts * y_p

    # extero sensor: position
    y_e = x[0,0] + np.random.randn()*sigma_e

    p_true_hist.append(x[0,0])
    p_hat_hist.append(p_hat_from_proprio)
    p_extero_hist.append(y_e)

print("Final true position:", p_true_hist[-1])
print("Final proprio-integrated estimate:", p_hat_hist[-1])
print("Final extero measurement:", p_extero_hist[-1])

File: Tutorials/Codes/Introduction_to_Robotics_Course/Chapter7/Lesson1/Chapter7_Lesson1.m
Ts = 0.1;
A = [1 Ts; 0 1];
B = [0.5*Ts^2; Ts];

x = [0;1];
b = 0.05; sigma_p = 0.02; sigma_e = 0.05;
p_hat = 0;

N = 200;
p_true = zeros(N,1);
p_hat_hist = zeros(N,1);
p_extero = zeros(N,1);

for k=1:N
    u = 0;
    x = A*x + B*u;

    y_p = x(2) + b + sigma_p*randn; % proprio (velocity)
    y_e = x(1) + sigma_e*randn;     % extero (position)

    p_hat = p_hat + Ts*y_p;

    p_true(k) = x(1);
    p_hat_hist(k) = p_hat;
    p_extero(k) = y_e;
end

disp([p_true(end), p_hat_hist(end), p_extero(end)])
